File: stack/linkedlist_stack.py
import linked_list

class LinkedListStack:
    """implement stack with linked list"""

    def __init__(self):
        """constructor"""
        self._peek: ListNode | None = None
        self._size: int = 0

    def size(self) -> int:
        """get the size of the stack"""
        return self._size

    def is_empty(self) -> bool:
        """whether the stack is empty"""
        return not self._peek

    def push(self, val: int):
        """push element into the stack"""
        node = ListNode(val)
        node.next = self._peek
        self._peek = node
        self._size += 1

    def pop(self) -> int:
        """pop element out of the stack"""
        num = self.peek()
        self._peek = self._peek.next
        self._size -= 1
        return num

    def peek(self) -> int:
        """get the element on the top"""
        if self.is_empty():
            raise IndexError("the stack is empty")
        return self._peek.val

    def to_list(self) -> list[int]:
        """convert the stack to list"""
        arr = []
        node = self._peek
        while node:
            arr.append(node.val)
            node = node.next
        arr.reverse()
        return arr

File: c/stack/linkedlist_stack.c
/*  implement stack with linked list */
LinkedListStack *newLinkedListStack() {
    LinkedListStack *s = malloc(sizeof(LinkedListStack));
    s->top = NULL;
    s->size = 0;
    return s;
}

/* destructor */
void delLinkedListStack(LinkedListStack *s) {
    while (s->top) {
        ListNode *n = s->top->next;
        free(s->top);
        s->top = n;
    }
    free(s);
}

/* get the size of stack */
int size(LinkedListStack *s) {
    assert(s);
    return s->size;
}

/* check if the stack is empty */
bool isEmpty(LinkedListStack *s) {
    assert(s);
    return size(s) == 0;
}

/* get the top element of stack */
int peek(LinkedListStack *s) {
    assert(s);
    assert(size(s) != 0);
    return s->top->val;
}

/* push element into stack */
void push(LinkedListStack *s, int num) {
    assert(s);
    ListNode *node = (ListNode *)malloc(sizeof(ListNode));
    node->next = s->top; // update next pointer
    node->val = num;     // update value
    s->top = node;       // update top pointer
    s->size++;           // update size
}

/* pop element out of stack */
int pop(LinkedListStack *s) {
    if (s->size == 0) {
        printf("stack is empty.\n");
        return INT_MAX;
    }
    assert(s);
    int val = peek(s);
    ListNode *tmp = s->top;
    s->top = s->top->next;
    // free memory
    free(tmp);
    s->size--;
    return val;
}
